File: php/src/detectors/PathTraversalDetector.php
<?php

namespace FireWTWall\Detectors;

class PathTraversalDetector
{
    private static array $rules = [
        ['name' => 'path-traversal-dotdot',  'severity' => 'critical', 'pattern' => '/(?:\.\.[\\/]|[\\/]\.\.)/'],
        ['name' => 'path-traversal-encoded', 'severity' => 'critical', 'pattern' => '/%2e%2e[%2f5c]/i'],
        ['name' => 'path-traversal-unicode', 'severity' => 'critical', 'pattern' => '/(?:%c0%ae|%c1%9c)/i'],
        ['name' => 'path-null-byte',         'severity' => 'critical', 'pattern' => '/%00|\x00/'],
        ['name' => 'path-etc-passwd',        'severity' => 'critical', 'pattern' => '/\/etc\/(?:passwd|shadow|hosts|group)\b/'],
        ['name' => 'path-win-system',        'severity' => 'critical', 'pattern' => '/(?:c:|%systemroot%)[\/\\\\]/i'],
        ['name' => 'path-env-file',          'severity' => 'high',     'pattern' => '/(?:^|\/)\.env(?:\.|$)/'],
        ['name' => 'path-wp-config',         'severity' => 'high',     'pattern' => '/wp-config\.php/i'],
        ['name' => 'path-htaccess',          'severity' => 'high',     'pattern' => '/\.htaccess\b/i'],
        ['name' => 'path-git-config',        'severity' => 'high',     'pattern' => '/\.git[\\/]/i'],
        ['name' => 'path-ssh-keys',          'severity' => 'high',     'pattern' => '/\.ssh[\\/]/i'],
        ['name' => 'path-proc-self',         'severity' => 'critical', 'pattern' => '/\/proc\/self\//i'],
        ['name' => 'path-php-wrappers',      'severity' => 'high',     'pattern' => '/(?:php|zip|phar|data|expect|glob|file):\/\//i'],
        ['name' => 'path-php-filter',        'severity' => 'high',     'pattern' => '/php:\/\/(?:filter|input|stdin)/i'],
    ];

    public static function scan(array $sources): ?array
    {
        foreach ($sources as $label => $values) {
            $result = self::scanValues($values, $label);
            if ($result !== null) return $result;
        }
        return null;
    }

    private static function scanValues($data, string $label): ?array
    {
        if (is_string($data)) {
            return self::matchString($data, $label);
        }
        if (is_array($data)) {
            foreach ($data as $value) {
                $r = self::scanValues($value, $label);
                if ($r !== null) return $r;
            }
        }
        return null;
    }

    private static function matchString(string $value, string $label): ?array
    {
        foreach (self::$rules as $rule) {
            if (preg_match($rule['pattern'], $value, $m)) {
                return [
                    'rule'     => $rule['name'],
                    'severity' => $rule['severity'],
                    'matched'  => substr($m[0], 0, 120),
                    'source'   => $label,
                ];
            }
        }
        return null;
    }
}

File: aspnet/src/detectors/PathTraversalDetector.cs
using System.Collections.Generic;
using System.Text.RegularExpressions;

namespace FireWTWall.Detectors
{
    public static class PathTraversalDetector
    {
        private static readonly (string Name, string Severity, Regex Pattern)[] Rules =
        {
            ("path-traversal-dotdot",  "critical", new Regex(@"(?:\.\.[\\/]|[\\/]\.\.)",                      RegexOptions.Compiled)),
            ("path-traversal-encoded", "critical", new Regex(@"%2e%2e[%2f5c]",                                RegexOptions.IgnoreCase | RegexOptions.Compiled)),
            ("path-traversal-unicode", "critical", new Regex(@"(?:%c0%ae|%c1%9c)",                            RegexOptions.IgnoreCase | RegexOptions.Compiled)),
            ("path-null-byte",         "critical", new Regex(@"%00|\x00",                                     RegexOptions.Compiled)),
            ("path-etc-passwd",        "critical", new Regex(@"/etc/(?:passwd|shadow|hosts|group)\b",         RegexOptions.Compiled)),
            ("path-win-system",        "critical", new Regex(@"(?:c:|%systemroot%)[/\\]",                     RegexOptions.IgnoreCase | RegexOptions.Compiled)),
            ("path-env-file",          "high",     new Regex(@"(?:^|/)\.env(?:\.|$)",                         RegexOptions.Compiled)),
            ("path-wp-config",         "high",     new Regex(@"wp-config\.php",                               RegexOptions.IgnoreCase | RegexOptions.Compiled)),
            ("path-htaccess",          "high",     new Regex(@"\.htaccess\b",                                 RegexOptions.IgnoreCase | RegexOptions.Compiled)),
            ("path-git-config",        "high",     new Regex(@"\.git[/\\]",                                   RegexOptions.IgnoreCase | RegexOptions.Compiled)),
            ("path-ssh-keys",          "high",     new Regex(@"\.ssh[/\\]",                                   RegexOptions.IgnoreCase | RegexOptions.Compiled)),
            ("path-proc-self",         "critical", new Regex(@"/proc/self/",                                  RegexOptions.IgnoreCase | RegexOptions.Compiled)),
            ("path-php-wrappers",      "high",     new Regex(@"(?:php|zip|phar|data|expect|glob|file)://",   RegexOptions.IgnoreCase | RegexOptions.Compiled)),
            ("path-php-filter",        "high",     new Regex(@"php://(?:filter|input|stdin)",                 RegexOptions.IgnoreCase | RegexOptions.Compiled)),
            ("path-windows-root",      "high",     new Regex(@"[a-zA-Z]:\\|%SYSTEMROOT%|%WINDIR%",           RegexOptions.IgnoreCase | RegexOptions.Compiled)),
            ("path-system32",          "critical", new Regex(@"windows[/\\]system32",                        RegexOptions.IgnoreCase | RegexOptions.Compiled)),
            ("path-boot",              "critical", new Regex(@"/boot/grub|/boot/vmlinuz|/boot/initrd",       RegexOptions.IgnoreCase | RegexOptions.Compiled)),
        };

        public static DetectorResult Scan(Dictionary<string, string> sources)
        {
            foreach (var kv in sources)
            {
                var r = MatchString(kv.Value, kv.Key);
                if (r != null) return r;
            }
            return null;
        }

        private static DetectorResult MatchString(string value, string label)
        {
            foreach (var rule in Rules)
            {
                Match m = rule.Pattern.Match(value);
                if (m.Success)
                    return new DetectorResult(rule.Name, rule.Severity,
                        m.Value.Length > 120 ? m.Value.Substring(0, 120) : m.Value, label);
            }
            return null;
        }
    }
}
